test_task_packed_picking/models/stock_picking.py

```diff
@@ -0,0 +1,112 @@
+        - Assign an owner
+        - Add products and set qty_done
+        - Mark as "Todo"
+        - Put in pack
+        Args:
+        operation_type (stock.picking.type): Operation type
+        stock_move_data (List of tuples): [(product_id, qty_done, serial)]
+        - (Integer) product_id: id of the product
+        - (float) qty_done: quantity done
+        - (Char, optional) serial: serial number to assign.
+        Default lot names will be used if is None or == False
+        Used only if 'create_lots==True'
+        owner (res.partner, optional): Owner of the product
+        location (stock.location, optional): Source location if differs from the
+        operation type one
+        location_dest (stock.location, optional): Destination location if differs
+        from the operation type one
+        package_name (Char, optional): Name to be assigned to the package. Default
+        name will be used if not provided.
+        set_ready (bool, optional): Try to set picking to the "Ready" state.
+        Returns:
+        stock.picking: Created picking
+        """
+        if not location:
+            location = operation_type.default_location_src_id
+        if not location_dest_id:
+            location_dest_id = operation_type.default_location_dest_id
+        if not owner:
+            owner = self.env.user.partner_id
+        vals = {
+            "picking_type_id": operation_type.id,
+            "location_id": location.id,
+            "location_dest_id": location_dest_id.id,
+            "owner_id": owner.id,
+        }
+        stock_picking_id = self.env["stock.picking"].create(vals)
+        for product_id, qty_done, serial in stock_move_data:
+            product = self.env["product.product"].browse(product_id)
+            stock_move_vals = {
+                "name": product.name,
+                "product_id": product.id,
+                "product_uom_qty": qty_done,
+                "quantity_done": qty_done,
+                "product_uom": product.uom_id.id,
+                "location_id": location.id,
+                "location_dest_id": location_dest_id.id,
+                "picking_id": stock_picking_id.id,
+            }
+            if create_lots and serial:
+                stock_move_vals.update(
+                    {
+                        "lot_ids": [
+                            (
+                                0,
+                                0,
+                                {
+                                    "name": serial,
+                                    "product_id": product.id,
+                                },
+                            )
+                        ]
+                    }
+                )
+            self.env["stock.move"].create(stock_move_vals)
+        stock_picking_id.action_confirm()
+        stock_picking_id.action_set_quantities_to_reservation()
+        if package_name:
+            package_id = self.env["stock.quant.package"].create(
```

Comment:
Split complex methods into sub functions for the  ease ot testing and extension. Check the [OCA Guidelines](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#idioms) for more details.
Especially when it comes the creation of new records